Add a reset button to individual keybinds

SleepySwords opened the pull request:
Closes #12457

SleepySwords committed:
Adding a reset button to individual keybinds

SleepySwords committed:
Rename Tests

bdach commented:
I'm not really a UX person but this colour choice seems bad. Green is not a colour I'd use for a destructive operation. Especially so given that the "reset all in section" button is also pink.

![2021-05-09-120233_545x130_scrot](https://user-images.githubusercontent.com/20418176/117567914-9aa96680-b0be-11eb-9995-31677f3e787e.png)

peppy commented:
Is there a reason this isn't being done using the `RestoreDefaultValueButton` implementation? I'd rather this uses an existing standard.

bdach commented:
If a binding has two associated key combination, the current implementation allows for resetting them individually. While using "the nub" we use anywhere else would probably to have to reset them both. Not that I mind changing to that behaviour.

Also if the nub is added, it might be a bit weird UX-wise that it'll be the only operation that can be performed on a binding without clicking on the keybind button. But again, not a UX person, so...

peppy commented:
I'd say resetting all and letting the user re-choose seems fine. The yellow indicator also has the added bonus of letting the user know which values they actually have changed

bdach reviewed (changes requested):
should use `RestoreDefaultValueButton` - see above discussion

HolaSwords commented:
I'm quite stuck on how to go about implementing the changes.

Should I make the KeyBindRow a SettingsItem? Or do I have to move (or edit the access modifier) of the RestoreDefaultValueButton?

peppy commented:
I think moving the sub-component out to be reused may work better. But it's not going to be a super-simple one either way, since we're not dealing with bindables here directly.

SleepySwords committed:
Well it works, just pretty ugly looking.

SleepySwords committed:
Fixed formatting

SleepySwords committed:
Merge branch 'master' into single-bind-reset-button

HolaSwords commented:
![Screen Shot 2021-05-15 at 1 08 48 am](https://user-images.githubusercontent.com/33922797/118290745-1de11700-b51a-11eb-86d4-af5f8047f844.png)

So while the latest commits do add the nub, there should be a better way in implementing this.

SleepySwords committed:
Reformatting

SleepySwords committed:
Merge branch 'single-bind-reset-button' of https://github.com/holaswords/osu into single-bind-reset-button

SleepySwords committed:
Reformatting

bdach commented on osu.Game.Tests/Visual/Settings/TestSceneKeyBindingPanel.cs:
```diff
@@ -71,7 +72,7 @@ public void TestClearButtonOnBindings()
             AddStep("click first row with two bindings", () =>
             {
-                multiBindingRow = panel.ChildrenOfType<KeyBindingRow>().First(row => row.Defaults.Count() > 1);
+                multiBindingRow = panel.ChildrenOfType<SettingsKeyBindingRow>().First(row => row.KeyBindingRow.Defaults.Count() > 1).KeyBindingRow;
```
Not sure what these changes are about. `ChildrenOfType()` is recursive, so it should continue to work the same way it did even with the component changes, I think.

HolaSwords commented on osu.Game/Overlays/KeyBinding/SettingsKeyBindingRow.cs:
```diff
@@ -0,0 +1,69 @@
+        public IEnumerable<string> FilterTerms => bindings.Select(b => b.KeyCombination.ReadableString()).Prepend(defaultGroup.Key.ToString());
+
+        public SettingsKeyBindingRow(
+            IGrouping<object, Framework.Input.Bindings.KeyBinding> defaultGroup,
```
I'm also using this to get the Defaults. Should I pass the Defaults separately?

HolaSwords commented on osu.Game/Overlays/KeyBinding/SettingsKeyBindingRow.cs:
```diff
@@ -0,0 +1,69 @@
+            this.defaultGroup = defaultGroup;
+            this.bindings = bindings;
+
+            KeyBindingRow = new KeyBindingRow(defaultGroup.Key, bindings.Where(b => ((int)b.Action).Equals((int)defaultGroup.Key)))
```
Perhaps a KeyBindingRow and BasicKeyBindingRow?
or
KeyBindingWithDefaultRestore and KeyBindingRow?

HolaSwords replied:
I ended doing RestorableKeyBindingRow and KeyBindingRow

SleepySwords committed:
Adding Requested changed

SleepySwords committed:
Formatting

SleepySwords committed:
Formatting

SleepySwords committed:
Removed unused method

SleepySwords committed:
Merge branch 'master' into single-bind-reset-button

SleepySwords committed:
Fixing tests

For some reason moving the mouse and clicking doesn't work with "dotnet test", but works when you run the osu.Game.Tests project.

SleepySwords committed:
Merge branch 'master' into single-bind-reset-button

SleepySwords committed:
Finishing requested changes, and tidy up

SleepySwords committed:
Fixes ResotreDefaultValue to use the BindableWithCurrent correctly

SleepySwords committed:
Merge branch 'master' into single-bind-reset-button

bdach committed:
Move value change bindings to `LoadComplete()`

Also removes a redundant one from the setter of `Current`. If the set to
current comes with an associated value change, the bind-unbind flow that
`BindableWithCurrent` implements will handle that change anyway.

bdach committed:
Remove now-redundant null check

bdach committed:
Simplify bindable flow in `KeyBindingRow`

bdach committed:
Revert to previous width spec instead of size

bdach commented:
I have started cleaning this up to hopefully get it to a mergeable state, but somewhat lost confidence that this is the correct way to go halfway in. The bindable bool stuff shoved into `RestoreValueDefaultButton` feels kinda bad to be honest. What would work really well here is a plain old action on click, without any of the bindable stuff. Design needs work, too:

![2021-05-17-212357_543x330_scrot](https://user-images.githubusercontent.com/20418176/118544912-33db1b80-b756-11eb-97ee-d7157b93facf.png)

(note how the rows are now significantly narrower than the button underneath them). And the naming is still pretty bad (`RestorableKeyBindingRow` doesn't only have restore functionality in it). But before I dig further I'd like a sanity check that the overall structure can be considered good enough to go or if it needs changing.

peppy commented:
For the design issue, I think this can easily be fixed by making the button less wide than its preceding elements:

![20210518 134639 (dotnet)](https://user-images.githubusercontent.com/191335/118591667-87457c00-b7df-11eb-841b-32352d543aee.png)

Will push this for now and review the rest of the changes.

peppy committed:
Update "reset all bindings" button to better match new key binding row widths

peppy committed:
Move private methods down

peppy commented on osu.Game/Overlays/KeyBinding/KeyBindingRow.cs:
```diff
@@ -118,6 +125,24 @@ private void load(OsuColour colours)
+
+            IsDefault.BindValueChanged(resetButtons =>
+            {
+                if (resetButtons.NewValue && !computeIsDefaultValue())
```
What is the `computeIsDefaultValue` call here for? Feels like this needs an inline comment if there's a reason for it to exist.

bdach replied:
This is the awkward part caused by the bool bindable stuff I talked about. In short, this bindable's can change to `true` in two cases:

* User clicks restore value button to restore binding
* User manually changes binding from non-default to default

This check is supposed to prevent the restore/finalise chain from running in case 2, while still keeping case 1 working. I suppose it's not strictly needed, because the restore would do nothing anyways, but that's also a potential argument for keeping it (to avoid busy work).

If the restore value button just fired a plain action this would not be necessary. The keybinding-button value synchronisation would still have to remain in some capacity.

peppy replied:
Right. I'm not sure how I feel about this being a bindable, since it looks like it's being updated at two very different locations. I'm fine with the bindable being used to convey *outwards* to the restore default button (ie. change the `public` exposed bindable to `IBindable`) but the actual action should probably be a separate action coming in.

HolaSwords replied:
I'm not sure what you mean by this? Doesn't the RestorableKeyBindingRow need a Bindable?

Also if you are just concerned with this awkward part and not that concerned about using a BindableBool in KeyBindingRow, we could just make the ResetButton action use the IsDefault property and make the RestoreDefaults method private (though that would make the code more unreadable) `k.KeyBindingRow.IsDefault.Value = true`

peppy replied:
Basically I'm saying the flow of information (whether it is default or not) is fine to be left using bindable flow, but the *set* operation should likely not be done using the same bindable.

Rationale is that it's set in two different places.

bdach replied:
What I was loosely thinking is that the default value button could have a publicly-declared `Action` attached, that in the 99% defaults to resetting `Current`, and would be overridden here to manually reset the row via a method call or otherwise. Haven't gotten around to trying it yet, though.

peppy replied:
@sleepyswords are you okay to implement this change?

SleepySwords replied:
Yeah sure, I'll make the Restore button implement an OsuButton,

Do you still want me to use an IBindable for the KeyBindingRow externally?

peppy commented on osu.Game/Overlays/KeyBinding/KeyBindingRow.cs:
```diff
@@ -51,13 +52,17 @@ public bool MatchingFilter
 
+        public Bindable<bool> IsDefault { get; } = new BindableBool(true)
+        {
+            Default = true
```
Unnecessary; this is handled by the ctor value.

peppy commented on osu.Game/Overlays/KeyBinding/RestorableKeyBindingRow.cs:
```diff
@@ -0,0 +1,71 @@
+
+        public IEnumerable<string> FilterTerms => bindings.Select(b => b.KeyCombination.ReadableString()).Prepend(key.ToString());
+
+        public RestorableKeyBindingRow(
```
We don't format like this anywhere

Joehuu replied:
Here is one example where that format is used.

https://github.com/ppy/osu/blob/44b8725458a80c57b9c38337a1b7c42a91f2d226/osu.Game/Online/API/Requests/SearchBeatmapSetsRequest.cs#L47-L60

bdach replied:
I glanced over this as I actually tend to format code similarly with hanging indents if there are a lot of ctor params (i.e. constructor DI), but I can change to single line.

peppy commented on osu.Game/Overlays/KeyBinding/RestorableKeyBindingRow.cs:
```diff
@@ -0,0 +1,71 @@
+            this.key = key;
+            this.bindings = bindings;
+
+            RestoreDefaultValueButton<bool> restoreDefaultButton;
```
Unnecessary; use

```csharp
Current = { BindTarget = KeyBindingRow.IsDefault }
```

peppy commented on osu.Game/Overlays/KeyBinding/RestorableKeyBindingRow.cs:
```diff
@@ -0,0 +1,71 @@
+            AutoSizeAxes = Axes.Y;
+            Padding = new MarginPadding { Right = SettingsPanel.CONTENT_MARGINS };
+
+            KeyBindingRow = new KeyBindingRow(key, bindings.Where(b => ((int)b.Action).Equals((int)key)))
```
This can also be inlined.

peppy replied:
Wasn't done.

Please avoid marking review comments resolved; we will do that on re-checking the PR.

peppy reviewed (changes requested):
General direction seems fine.

peppy committed:
Remove outdated comment

SleepySwords committed:
RestoreDefaultValueButton implements OsuButton

SleepySwords commented:
Is there a way we are able to combine the RestorableKeyBindingRow (need to change the name if it stays the same) and the KeyBindingRow?

SleepySwords committed:
Formatting

SleepySwords committed:
Renaming files

SleepySwords committed:
Apply the rest of requested changes

SleepySwords committed:
Restore class names

SleepySwords committed:
Formatting, renaming

SleepySwords committed:
Removing redundant argument list

SleepySwords committed:
Fixing errors

peppy committed:
Merge branch 'master' into single-bind-reset-button

peppy committed:
Replace local namespace qualifiers with `using`

peppy committed:
Inline `KeyBindingRow` construction

peppy committed:
Add missing newline

peppy committed:
Combine `RestorableKeyBindingRow` back into `KeyBindingRow`

peppy commented:
So... I've combined the weird `RestorableKeyBindingRow` class with the old one. To make this work, I had to add one workaround to the `RestoreDefaultsButton`, which is to have it handling focus. If this isn't done, the focus request will propagate upwards and be handled by the `KeyBindingRow` (this is framework behaviour).

@smoogipoo please review this one for me and make sure you're okay with the above.

I did dabble with moving all the focus to the `KeyButton` itself, which seems like the more appropriate place to handle it, but it complicates the event flow when the user clicks between buttons in the same row. Might be something we consider in  the future if refactoring this component further, but I'm just looking to keep changes to a minimum for now (which is why I combined the classes in the first place).

peppy committed:
Simplify default handling flow

smoogipoo reviewed (approved):
lgtm

smoogipoo committed:
Merge branch 'master' into single-bind-reset-button
